File: packages/protocol/src/merkle.ts
/**
 * `sha256-merkle-sorted-v1` — the tree a kind 8108 checkpoint commits to.
 *
 * A checkpoint is a relay signing "this is the set of event ids I hold for this
 * group in this window". The set has to be reducible to 32 bytes that anyone
 * holding the same events can recompute, and the reduction has to be specified
 * tightly enough that a Go relay and a TypeScript client never disagree about
 * it — a root that differs by one byte is indistinguishable from a relay caught
 * lying, which is the worst possible failure mode for a mechanism whose entire
 * output is an accusation.
 *
 * So the rules are few and all of them are load-bearing:
 *
 * **Sorted and deduplicated.** The relay's storage order is its own business
 * and a client receives events in whatever order a subscription delivers them.
 * Sorting is what makes the root a function of the *set*. Ids are compared as
 * lowercase hex strings, which is the same ordering as on the underlying bytes.
 *
 * **Leaves and internal nodes are hashed with different prefixes** — `0x00` for
 * a leaf, `0x01` for a pair. Without that separation an internal node can be
 * presented as a leaf: an attacker who controls two ids can offer their
 * concatenation as a single "event id" and produce a valid-looking inclusion
 * proof for an event nobody published. This is RFC 6962's construction and the
 * reason is the same there.
 *
 * **An odd node is promoted, never duplicated.** Duplicating the last node to
 * pad a level is the Bitcoin construction, and it is broken: a set of three
 * leaves `[a, b, c]` and a set of four `[a, b, c, c]` produce the same root, so
 * a relay could commit to one set and later claim it meant the other. Promotion
 * makes the tree shape a function of the leaf count alone.
 *
 * **The empty set is `sha256("")`.** A quiet window is a real case — a
 * workspace overnight — and the checkpoint chain must not break for it. Picking
 * an arbitrary sentinel like 32 zero bytes would be equally valid and less
 * checkable; this one is a value anyone can reproduce from the name.
 */

import { sha256 } from '@noble/hashes/sha2.js'
import { bytesToHex, hexToBytes } from '@noble/hashes/utils.js'

/** The algorithm name carried in a checkpoint body, so a verifier never guesses. */
export const MERKLE_ALGORITHM = 'sha256-merkle-sorted-v1'

/** `sha256("")` — the root of an empty window. */
export const EMPTY_ROOT = bytesToHex(sha256(new Uint8Array(0)))

const LEAF = 0x00
const NODE = 0x01

/**
 * One step of an inclusion proof: a sibling hash and which side it sat on.
 *
 * A promoted node contributes no step, which is why a path can be shorter than
 * `ceil(log2(n))` and why the verifier must not assume a length.
 */
export interface MerkleStep {
  hash: string
  side: 'left' | 'right'
}

/** Sorted, deduplicated, lowercased leaves — the canonical form of a set of ids. */
export function merkleLeaves(ids: readonly string[]): string[] {
  return [...new Set(ids.map((id) => id.toLowerCase()))].sort()
}

/**
 * The root over a set of event ids.
 *
 * Order and duplicates in `ids` are irrelevant by construction; pass whatever a
 * query returned.
 */
export function merkleRoot(ids: readonly string[]): string {
  const leaves = merkleLeaves(ids)
  if (leaves.length === 0) return EMPTY_ROOT

  let level = leaves.map((id) => hashLeaf(id))
  while (level.length > 1) {
    const next: Uint8Array[] = []
    for (let i = 0; i < level.length; i += 2) {
      // The odd one out moves up untouched. Hashing it with itself would make
      // [a,b,c] and [a,b,c,c] the same tree.
      next.push(i + 1 < level.length ? hashNode(level[i]!, level[i + 1]!) : level[i]!)
    }
    level = next
  }
  return bytesToHex(level[0]!)
}

/**
 * The audit path proving `id` is one of `ids`, or `undefined` if it is not.
 *
 * The point of a path is that a client holding one event can check it against a
 * committed root without refetching the window — O(log n) hashes instead of
 * however many events a busy day produced.
 */
export function merkleProof(ids: readonly string[], id: string): MerkleStep[] | undefined {
  const leaves = merkleLeaves(ids)
  let index = leaves.indexOf(id.toLowerCase())
  if (index === -1) return undefined

  const path: MerkleStep[] = []
  let level = leaves.map((leaf) => hashLeaf(leaf))
  while (level.length > 1) {
    const next: Uint8Array[] = []
    for (let i = 0; i < level.length; i += 2) {
      if (i + 1 >= level.length) {
        next.push(level[i]!)
        continue
      }
      if (i === index) path.push({ hash: bytesToHex(level[i + 1]!), side: 'right' })
      else if (i + 1 === index) path.push({ hash: bytesToHex(level[i]!), side: 'left' })
      next.push(hashNode(level[i]!, level[i + 1]!))
    }
    index = Math.floor(index / 2)
    level = next
  }
  return path
}

/** True if `path` carries `id` up to `root`. */
export function verifyMerkleProof(id: string, path: readonly MerkleStep[], root: string): boolean {
  let hash = hashLeaf(id.toLowerCase())
  for (const step of path) {
    const sibling = hexToBytes(step.hash)
    hash = step.side === 'left' ? hashNode(sibling, hash) : hashNode(hash, sibling)
  }
  return bytesToHex(hash) === root.toLowerCase()
}

function hashLeaf(id: string): Uint8Array {
  return sha256(prefixed(LEAF, hexToBytes(id)))
}

function hashNode(left: Uint8Array, right: Uint8Array): Uint8Array {
  return sha256(prefixed(NODE, left, right))
}

function prefixed(tag: number, ...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 1)
  const out = new Uint8Array(total)
  out[0] = tag
  let at = 1
  for (const part of parts) {
    out.set(part, at)
    at += part.length
  }
  return out
}

File: apps/relay/internal/checkpoint/merkle.go
// Package checkpoint implements ordering integrity layer 3: the relay signing a
// commitment to the set of events it holds for a group in a window.
//
// Layers 1 and 2 — per-author `counter` tags and causal `e` tags — let a reader
// notice gaps in what it already holds. Neither can see an event that was never
// served, and a relay quietly serving a smaller world is not detectable from
// NIP-01 at all. That is by design: nothing in Nostr obliges a relay to admit
// what it has.
//
// A checkpoint does not change that obligation; it makes the relay's own claim
// checkable. Having signed "these are the ids I held for this group up to T",
// the relay cannot later serve a set missing one of them without producing, in
// the reader's hands, a signed contradiction. The honest limit, stated here
// because it belongs beside the mechanism rather than in a footnote: a relay
// that publishes no checkpoints is not caught by any of this. Its silence is at
// least visible, which silent withholding is not.
//
// This file is the tree. `checkpoint.go` is the thing that cuts windows and
// signs.
package checkpoint

import (
	"crypto/sha256"
	"encoding/hex"
	"sort"
	"strings"
)

// Algorithm is the name a checkpoint body carries, so a verifier never guesses.
const Algorithm = "sha256-merkle-sorted-v1"

const (
	leafTag = 0x00
	nodeTag = 0x01
)

// EmptyRoot is sha256(""), the root of a window in which nothing happened.
//
// A quiet window is a real case — a workspace overnight — and the chain must
// not break for it. An arbitrary sentinel like 32 zero bytes would be equally
// valid and less checkable; this one anybody can reproduce from the name.
var EmptyRoot = hex.EncodeToString(sha256.New().Sum(nil))

// Step is one hop of an inclusion proof: a sibling hash and the side it sat on.
type Step struct {
	Hash string `json:"hash"`
	Side string `json:"side"` // "left" or "right"
}

// Leaves is the canonical form of a set of ids: sorted, deduplicated, lowercase.
//
// Sorting is what makes the root a function of the *set* rather than of the
// order a query happened to return. Deduplicating matters because a relay may
// legitimately serve one event twice across overlapping filters, and a reader
// recomputing the root from what it received must land in the same place.
func Leaves(ids []string) []string {
	seen := make(map[string]struct{}, len(ids))
	out := make([]string, 0, len(ids))
	for _, id := range ids {
		lower := strings.ToLower(id)
		if _, dup := seen[lower]; dup {
			continue
		}
		seen[lower] = struct{}{}
		out = append(out, lower)
	}
	sort.Strings(out)
	return out
}

// Root reduces a set of event ids to the 32 bytes a checkpoint commits to.
//
// Two rules here are load-bearing and neither is the obvious choice:
//
// Leaves and internal nodes are hashed under different prefixes (RFC 6962's
// construction). Without that separation an internal node can be presented as a
// leaf — an attacker who controls two ids offers their parent as a single
// "event id" and produces a valid-looking inclusion proof for an event nobody
// published.
//
// An odd node is promoted, never duplicated. Padding a short level by repeating
// its last node is Bitcoin's rule and it is broken (CVE-2012-2459): [a,b,c] and
// [a,b,c,c] produce the same root, so a relay could commit to one set and later
// insist it meant the other, with one signature backing both stories.
func Root(ids []string) string {
	leaves := Leaves(ids)
	if len(leaves) == 0 {
		return EmptyRoot
	}

	level := make([][]byte, len(leaves))
	for i, leaf := range leaves {
		level[i] = hashLeaf(leaf)
	}

	for len(level) > 1 {
		next := make([][]byte, 0, (len(level)+1)/2)
		for i := 0; i < len(level); i += 2 {
			if i+1 < len(level) {
				next = append(next, hashNode(level[i], level[i+1]))
			} else {
				next = append(next, level[i]) // promoted, untouched
			}
		}
		level = next
	}
	return hex.EncodeToString(level[0])
}

// Proof is the audit path showing id is one of ids, or nil with ok=false.
//
// The point is that a client holding one event can check it against a committed
// root without refetching the window — O(log n) hashes rather than however many
// events a busy day produced.
//
// A promoted node contributes no step, so a path can be shorter than
// ceil(log2(n)). A verifier that derives the expected length from the leaf count
// rather than walking what it was given passes every power-of-two case and
// fails on the first odd one.
func Proof(ids []string, id string) ([]Step, bool) {
	leaves := Leaves(ids)
	index := sort.SearchStrings(leaves, strings.ToLower(id))
	if index == len(leaves) || leaves[index] != strings.ToLower(id) {
		return nil, false
	}

	path := []Step{}
	level := make([][]byte, len(leaves))
	for i, leaf := range leaves {
		level[i] = hashLeaf(leaf)
	}

	for len(level) > 1 {
		next := make([][]byte, 0, (len(level)+1)/2)
		for i := 0; i < len(level); i += 2 {
			if i+1 >= len(level) {
				next = append(next, level[i])
				continue
			}
			switch index {
			case i:
				path = append(path, Step{Hash: hex.EncodeToString(level[i+1]), Side: "right"})
			case i + 1:
				path = append(path, Step{Hash: hex.EncodeToString(level[i]), Side: "left"})
			}
			next = append(next, hashNode(level[i], level[i+1]))
		}
		index /= 2
		level = next
	}
	return path, true
}

// VerifyProof reports whether path carries id up to root.
func VerifyProof(id string, path []Step, root string) bool {
	hash := hashLeaf(strings.ToLower(id))
	for _, step := range path {
		sibling, err := hex.DecodeString(step.Hash)
		if err != nil {
			return false
		}
		if step.Side == "left" {
			hash = hashNode(sibling, hash)
		} else {
			hash = hashNode(hash, sibling)
		}
	}
	return hex.EncodeToString(hash) == strings.ToLower(root)
}

func hashLeaf(id string) []byte {
	raw, err := hex.DecodeString(id)
	if err != nil {
		// An id that is not hex cannot be a Nostr event id. Hashing the bytes
		// as given keeps the function total rather than panicking on input a
		// caller should never have produced; it will simply not match anything.
		raw = []byte(id)
	}
	return prefixed(leafTag, raw)
}

func hashNode(left, right []byte) []byte {
	return prefixed(nodeTag, left, right)
}

func prefixed(tag byte, parts ...[]byte) []byte {
	hasher := sha256.New()
	hasher.Write([]byte{tag})
	for _, part := range parts {
		hasher.Write(part)
	}
	return hasher.Sum(nil)
}
